File: ruby/collatz-conjecture/collatz_conjecture.rb
# CollatzConjecture class
class CollatzConjecture
  def self.steps(num, count = 0)
    raise ArgumentError if num < 1
    return count if num == 1

    count += 1
    number = num.even? ? num / 2 : (3 * num) + 1
    steps(number, count)
  end
end

File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n = if n <= 0 then Nothing else step 0 n

step :: Integer -> Integer -> Maybe Integer
step count n
  | n == 1     = Just count
  | even n     = step (count + 1) (n `div` 2)
  | otherwise  = step (count + 1) (3 * n + 1)
